pkg/paths/path.go

```diff
@@ -50,3 +52,28 @@ func GetAllParams(r *http.Request) (*AllPathParams, error) {
 
 	return &AllPathParams{Module: mod, Version: version}, nil
 }
+
+// MatchGlobPattern reports whether any path prefix of name
+// matches the pattern (as defined by path.Match)
+func MatchGlobPattern(pattern, name string) (bool, error) {
+	const op errors.Op = "paths.MatchGlobPattern"
+	pCount := strings.Count(pattern, "/")
+	nCount := strings.Count(name, "/")
+
+	for nCount > pCount {
+		i := strings.LastIndex(name, "/")
+		if i == -1 {
+			break
+		}
+		name = name[:i]
+		nCount = strings.Count(name, "/")
```

Comment:
why not keep the same implementation as `matchesPattern` above? That one (copied from the Go source code) has one call to strings.Count while this implementation has N+2 calls?

Reply:
Good question. I actually wrote this several weeks ago; before the `matchesPattern` implementation (I just received permission for the postback).  It was written to be understandable, and not too much copy-paste from the Go source code.

After rebasing and seeing the new `matchesPattern` function, I benchmarked it. `MatchGlobPattern` runs ~4-7% faster than `matchesPattern`, so I left it with my new implementation.

```
name                        old time/op    new time/op    delta
PathMatch/MatchPattern/1-8     131ns ± 0%     125ns ± 0%  -4.58%
PathMatch/MatchPattern/2-8     148ns ± 0%     138ns ± 0%  -6.76%
PathMatch/MatchPattern/3-8     168ns ± 0%     160ns ± 0%  -4.76%
PathMatch/MatchPattern/4-8     200ns ± 0%     193ns ± 0%  -3.50%
PathMatch/MatchPattern/5-8     215ns ± 0%     207ns ± 0%  -3.72%
PathMatch/MatchPattern/6-8     236ns ± 0%     227ns ± 0%  -3.81%
PathMatch/MatchPattern/7-8     270ns ± 0%     249ns ± 0%  -7.78%
PathMatch/MatchPattern/8-8     290ns ± 0%     272ns ± 0%  -6.21%
PathMatch/MatchPattern/9-8     313ns ± 0%     292ns ± 0%  -6.71%
```

> The `/`+digit for each benchmark is the number of `/` in the pattern.

Reply:
Thanks for the stats. It definitely seems weird that more strings.Count calls is more performant.

I ran some benchmarks against the two functions with the following input

`MatchGlobPattern("github.com/one/two/*", "github.com/one/two/three/four/five/six/seven")`

and 

`matchesPattern("github.com/one/two/*", "github.com/one/two/three/four/five/six/seven")`

And the first one yielded `167 ns/op` while the second one yielded `115 ns/op` which makes a little more sense. 

I'm a bit curious as to how you got your stats, do you mind putting them in a gist? Just so I can see what the results are on my machine. 

Interestingly enough, `benchstat` returned this: 

```
name   old time/op    new time/op    delta
Match-8     167ns ± 0%     115ns ± 0%   ~     (p=1.000 n=1+1)
```

Where the `~` indicates that the difference is not significant enough :) 

That said, I think it's safe to keep the Go implementation because they have far more robust tests on their implementations and it's good to keep the code consistent.

Reply:
Ahh, looks like I made a mistake on my benchmark. The number of slashes between the pattern and the target were always matching. 

I can switch out my function for `matchesPattern`.

Reply:
So my original benchmark was testing increasing lengths of `pattern`, but it's much more likely that `pattern` will be short, and `target` long. So the `matchesPattern` implementation is much better, as it runs in constant time, even with increasing `target` length.

I've added a Benchmark so this can be monitored / compared in the future if needed.